### PR TITLE
Add ability to rebind to keyring passphrases and Clevis

### LINKED ISSUE #2358: Support changing passphrase via key description and clevis-luks-regen
While @mulkieran brought up the question of Tang key rotation for our Clevis integration and this process does appear to be manual, I do want to look into the details of handling key rotation. There is some information that indicates that newer Clevis clients automatically pick up new key advertisements, but we should verify this with the Clevis maintainers and determine whether the same keyslot will be used and overwritten as automatically binding to a new keyslot and leaving the old keyslot tied to the old keys could be a problem.

### PR DESCRIPTION
**Pull request by jbaublitz**
Closes #2358 
Supersedes #2604

### CODE REVIEW

**Comment by mulkieran** [comment]
@jbaublitz Please rebase here when convenient.

**Comment by jbaublitz** [comment]
@mulkieran Rebased.

**Comment by mulkieran** [comment]
@jbaublitz Could you rebase again. The rebase must have occurred just when GitHub Actions started to have trouble.

**Comment by jbaublitz** [comment]
@mulkieran I've rebased and I'll put it back in review once tests have passed.

**Comment by jbaublitz** [comment]
@mulkieran Back in review.

**Comment by mulkieran** [comment]
@jbaublitz Please rebase when you get the chance.

**Review by mulkieran — CHANGES_REQUESTED** [review]
Code looks great...I have two requests for additional comments and one request for comment.

In the stratis-cli PR https://github.com/stratis-storage/stratis-cli/pull/800 I assume that the user will only want to rebind w/ tang and the man page reflects this. But in this PR, the rebinding occurs with whatever Clevis configuration is present. Should we decide to:
1. Make the CLI less specific?
2. Return an error if the configuration is not for tang?

We allow any sort of configuration to be passed when binding via the API. So it seems like the API should be equally general when rebinding. However, I can not persuade myself that rebinding with tpm2 makes any sense, so that the call to "clevis luks regen" should always return an error in that case. If it does return an error then stratisd will return an error on the D-Bus and that will be, up to a point (module the ugliness of clevis error strings), the right behavior.

Anyhow, I've asked for your review on https://github.com/stratis-storage/stratis-cli/pull/800 in case you think the CLI side should be less specific.

But I'm not sure that what we have here isn't the best choice available; the cleanest and most consistent solution.

**Review by mulkieran — CHANGES_REQUESTED** [review]
Additional changes look great!

Could you grab the commit from here: https://github.com/stratis-storage/stratisd/pull/2682/commits/ef7ca83e6a3454a8fb0954dd88016c6df732289e.

Also, could you confirm that you would expect the thumbprint displayed by the stratisd report method to be the same before and after a rebind using tang? The theory is that the thumbprint is associated with the identification of the tang server, not with the particular keys that the tang server is publishing and so should be expected to remain constant while the tang server itself remains constant.

File: src/engine/strat_engine/backstore/crypt/shared.rs [hunk 1]
```diff
@@ -122,36 +122,53 @@ pub fn acquire_crypt_device(physical_path: &Path) -> StratisResult<CryptDevice>
...
+            log_on_failure!(
+                device.keyslot_handle().change_by_passphrase(
+                    None,
+                    None,
+                    pass.as_ref(),
+                    key.as_ref()
+                ),
+                "Failed to change passphrase for encrypted device"
+            ) as c_uint
+            // The above cast is a work around for bug in libcryptsetup-rs
```
[the remark this answers is not in the document]
Comment (jbaublitz) [reply src/engine/strat_engine/backstore/crypt/shared.rs]: Resolved.

File: src/engine/strat_engine/backstore/blockdevmgr.rs [hunk 2]
```diff
@@ -584,6 +587,96 @@ impl BlockDevMgr {
...
+        )?;
+
+        Ok(Some(true))
+    }
+
+    /// Regenerate the Clevis bindings with the block devices in this pool using
+    /// the same configuration.
+    ///
+    /// The method for this rollback caches the initial Clevis metadata and
+    /// reverts all of the devices if there is a failure.
```
[the remark this answers is not in the document]
Comment (jbaublitz) [reply src/engine/strat_engine/backstore/blockdevmgr.rs:621]: Resolved.

**Comment by jbaublitz** [comment]
> Additional changes look great!
> 
> Could you grab the commit from here: [ef7ca83](https://github.com/stratis-storage/stratisd/commit/ef7ca83e6a3454a8fb0954dd88016c6df732289e).
> 
> Also, could you confirm that you would expect the thumbprint displayed by the stratisd report method to be the same before and after a rebind using tang? The theory is that the thumbprint is associated with the identification of the tang server, not with the particular keys that the tang server is publishing and so should be expected to remain constant while the tang server itself remains constant.

This is now resolved. I didn't properly update the cached data structure after performing the action!

**Review by mulkieran — COMMENTED** [review]
Ok! This looks ready for testing again!

**Comment by jbaublitz** [comment]
@drckeefe @mulkieran I've confirmed again that removing the old Tang key after calling `RebindClevis` and then issuing `stratis pool unlock clevis` successfully unlocks the device. The issue did appear to be with the custom build of Clevis in my `PATH`.

**Comment by jbaublitz** [comment]
@drckeefe Rollback bug is fixed.

**Review by drckeefe — APPROVED** [review]

**Review by mulkieran — CHANGES_REQUESTED** [review]
This is ready to go. Please close and open a new PR for final review.

**Pull request closed without merging**